middleware/kubernetes/nametemplate/nametemplate.go

```diff
@@ -0,0 +1,177 @@
+package nametemplate
+
+import (
+	"errors"
+    "fmt"
+    "strings"
+
+	"github.com/miekg/coredns/middleware/kubernetes/util"
+)
+
+// Likely symbols that require support:
+// ${id}
+// ${ip}
+// ${portname}
+// ${protocolname}
+// ${servicename}
+// ${namespace}
+// ${type}              "svc" or "pod"
+// ${zone}
+
+
+// SkyDNS normal services have an A-record of the form "${servicename}.${namespace}.${type}.${zone}"
+// This resolves to the cluster IP of the service.
+
+// SkyDNS headless services have an A-record of the form "${servicename}.${namespace}.${type}.${zone}"
+// This resolves to the set of IPs of the pods selected by the Service. Clients are expected to
+// consume the set or else use round-robin selection from the set.
+
+
+var symbols = map[string]string{
+    "service": "${service}",
+    "namespace": "${namespace}",
+    "type": "${type}",
+    "zone": "${zone}",
+}
+
+var types = []string{
+	"svc",
+	"pod",
+}
+
+
+// TODO: Validate that provided NameTemplate string only contains:
+//			* valid, known symbols, or
+//			* static strings
+
+// TODO: Support collapsing multiple segments into a symbol. Either:
+//			* all left-over segments are used as the "service" name, or
+//			* some scheme like "${namespace}.${namespace}" means use
+//			  segments concatenated with a "." for the namespace, or
+//			* ${namespace2:4} means use segements 2->4 for the namespace.
```

Comment:
k8s doesn't need it. SkyDNS when used with k8s has defined a specific record name format. The introduction of the "svc" or "pod" subzone is part of the current SkyDNS+k8s record name format. To get CoreDNS+k8s to be a drop-in replacement for SkyDNS I am aiming to match the SkyDNS+k8s naming scheme.

Using a template to allow variation of the record name format is not necessary. It may be something that I drop and go back to a fixed record name format. However, I think templating the record name has some value if it can be handled without significant code complexity. The test will be whether I can implement wildcard lookups without the template support getting in the way.